---
title: Analysis of scRNA-seq data from CML patient 0114 (Jan 14, 2019)
author: Stephanie Hicks
output: 
    html_document:
        theme: cosmo 
        toc: true
        toc_float: true
        highlight: tango
        number_sections: false
fig_width: 5
fig_height: 5 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Summary of patient data

This is scRNA-seq (10X Genomics platform) data from a MPN patient 
(denoted `0114` patient) from Jan 14, 2019. The data is available 
on the JHPCE cluster and the following files are in the folder: 

- `barcodes_2.tsv`: barcode sequences of the detected cells.
- `features.tsv`: names of the genes (columns of the count matrix)
- `mycounts_2.csv`: count matrix (445.8 MB in size). The rows are the cells and the columns are the genes (6598 cells, 33694 genes). Each entry is the number of detected transcripts of a given gene in a given cell.
- `mygroupings_threshold_NS_100_Rev2.csv`: contains the number of wild-type or cancer transcripts detected in each cell. The columns are the cells (same number as the rows of the count matrix). There are two rows. Row 1 is the number of WT Jak2 transcripts. Row 2 is the number of cancer Jak2 transcripts.

The count matrix above is from our first sequencing run on this patient library. However, it has since been resequenced to increase depth by a factor of 3. This higher-depth data should be available by our meeting in Boston on March 8, 2019. 

## Previous analyses reported by Sahand

"We use SPRING, a dimensionality reduction 
tool that uses force-directed graph to plot the
differentiation trajectories. SPRING has some
criteria for selecting a subset of genes that are 
informative (for example fano factor and mean expression 
thresholds)."

# Biological question of interest

* Interested in identifying genes that are over-expressed or under-expressed in the cancer cells.
* Are there differences between cancer and normal cells in genes that regulate cell cycle?
* Hypothesis: Previous analyses suggested the gene E2F4 is over expressed in the cancer cells.

# Data Analysis

Questions for Sahand: 

* How to call whether a cell is a mutant or wildtype based on number of reads? 
* Would expect cells that have at least 1 wildtype or mutant read to have more 
UMI counts or more genes expressed? 

To-dos: 

* identify doublets, empty cells, and and apply normalization
* Use `scran::cyclone()` in Bioconductor to compare cancer vs normal cells in different stages of cell cycle


## Load data
Read in data 
```{r load-data, message=FALSE, cache=TRUE, eval=FALSE}
dat_counts <- readr::read_csv("data/0114P_RNAseq/BM_CD34/mycounts_2.csv",
                              col_names = FALSE) # dat is 445.8 MB in size
dat_counts <- t(dat_counts) # flip to make cells along columns

dat_muts <- readr::read_csv(
  "data/0114P_RNAseq/BM_CD34/mygroupings_threshold_NS_100_Rev2.csv", 
  col_names = FALSE)
dat_features <- readr::read_tsv(
  "data/0114P_RNAseq/BM_CD34/features.tsv", 
  col_names = FALSE)
dat_barcodes <- readr::read_tsv(
  "data/0114P_RNAseq/BM_CD34/barcodes_2.tsv", 
  col_names = FALSE)
```

Create a `SingleCellExperiment` object
```{r create-sce-object, message=FALSE, cache=TRUE, eval=FALSE}
library(SingleCellExperiment)

# row meta data 
sce_rowData <- as.data.frame(dat_features[,1:2])
colnames(sce_rowData) <- c("gene_name_ensembl", 
                           "gene_name_symbol")
rownames(sce_rowData) <- sce_rowData$gene_name_ensembl

# column meta data 
sce_colData <- data.frame("jak2_reads_wt" = as.numeric(dat_muts[1, -1]), 
                          "jak2_reads_cancer" = as.numeric(dat_muts[2, -1]))
rownames(sce_colData) <- dat_barcodes$X1

# counts matrix 
rownames(dat_counts) <- sce_rowData$gene_name_ensembl
colnames(dat_counts) <- dat_barcodes$X1

sce <- SingleCellExperiment(assays = list(counts = dat_counts), 
          rowData = sce_rowData, colData = sce_colData)
```

## Quality control 

```{r scater-qc-basic, message=FALSE, cache=TRUE, eval=FALSE}
library(scater)

sce <- getBMFeatureAnnos(sce, 
            ids = rownames(sce),
            filters = "ensembl_gene_id", 
            attributes = c("ensembl_gene_id", "hgnc_symbol",
                           "chromosome_name", "gene_biotype",
                           "start_position", "end_position"),
            biomart = "ENSEMBL_MART_ENSEMBL", 
            dataset = "hsapiens_gene_ensembl",
            host = "www.ensembl.org")

# separate information about mitochondrial genes
rowData(sce)[grep("^MT-", rowData(sce)$gene_name_symbol),]
isSpike(sce, "MT") <- grepl("^MT-", rowData(sce)$gene_name_symbol)

# calculate QC metrics
sce <- calculateQCMetrics(sce)

# save SCE object
saveRDS(sce, file = "data/0114P_RNAseq/BM_CD34/sce_0114.rds")
```


### Exploratory data analysis on cells 

```{r}
library(SingleCellExperiment)
library(scater)
sce <- readRDS("data/0114P_RNAseq/BM_CD34/sce_0114.rds")
```


Histogram of library sizes and number 
of expressed genes.
```{r}
par(mfrow=c(1,2))
hist(sce$total_counts/1e4, xlab="Library sizes (thousands)", main="",
     breaks=20, col="grey80", ylab="Number of cells")
hist(sce$total_features_by_counts, xlab="Number of expressed genes", main="",
     breaks=20, col="grey80", ylab="Number of cells")
```

Here is a table of the number of wild-type or cancer transcripts
detected in each cell covering Jak2. There are e.g. 11 cells that 
both have 1 read from the wildtype and cancer transcripts. Not sure 
how to handle those situations? 
```{r}
table(sce$jak2_reads_cancer, sce$jak2_reads_wt)
```

For now let's just assume if you have at least one cancer read 
covering Jak2, it is called a mutation. 

```{r}
sce$cancer_status <- ifelse(
  (sce$jak2_reads_cancer > 0 & sce$jak2_reads_wt < 1), "mut", 
  ifelse((sce$jak2_reads_cancer < 1 & sce$jak2_reads_wt > 0), "wt", "unknown" ))
sce$cancer_status <- factor(sce$cancer_status, levels=c("wt", "mut", "unknown"))
```

Let's look at the distribution of total counts and number 
of expressed genes when stratify by the number of cancer 
reads or wild type reads. Also what happens when you stratify 
by cancer status

```{r}
par(mfrow=c(2,3))
boxplot(sce$total_counts ~ sce$jak2_reads_cancer, log="y", 
        main = "jak2_reads_cancer", ylab = "total_counts" )
boxplot(sce$total_counts ~ sce$jak2_reads_wt, log="y", 
        main = "jak2_read_wt", ylab = "total_counts")
boxplot(sce$total_counts ~ sce$cancer_status, log="y", 
        main = "cancer_status", ylab = "total_counts")

boxplot(sce$total_features_by_counts ~ sce$jak2_reads_cancer, log="y", 
        main = "jak2_reads_cancer", ylab = "total_features_by_counts" )
boxplot(sce$total_features_by_counts ~ sce$jak2_reads_wt, log="y", 
        main = "jak2_read_wt", ylab = "total_features_by_counts")
boxplot(sce$total_features_by_counts ~ sce$cancer_status, log="y", 
        main = "cancer_status", ylab = "total_features_by_counts")
```

We should also look at the percent of reads mapping to MT genes, 
which could be evidence that we should exclude these cells

```{r}
par(mfrow=c(1,1))
boxplot(sce$pct_counts_MT ~ sce$cancer_status, ylim = c(0, 50), 
        ylab = "pct_counts_MT", main = "cancer_status")
```
There are definitely cells that have much higher pct MT
in the unknown cancer status group. 


### Remove low-quality cells 

We remove cells with log-library sizes that are more than 
3 median absolute deviations (MADs) below the median 
log-library size. (A log-transformation improves resolution 
at small values, especially when the MAD of the raw values 
is comparable to or greater than the median). Next, we 
remove cells where the log-transformed number of expressed 
genes is 3 MADs below the median. Then, we remove cells 
with percent mitochondria 3 MADs above the median. 

```{r}
libsize.drop <- isOutlier(sce$total_counts, nmads=3, type="lower", log=TRUE)
feature.drop <- isOutlier(sce$total_features_by_counts, nmads=3, type="lower", log=TRUE)
MT.drop <- isOutlier(sce$pct_counts_MT, nmads=3, type="higher", log=TRUE)

table(libsize.drop) 
table(feature.drop, MT.drop)

sce <- sce[,!(libsize.drop | feature.drop | MT.drop)]

# sanity check
par(mfrow=c(1,1))
plot(sce$total_features_by_counts, sce$pct_counts_MT)

```


