---
title: Analysis of Single Cell Dynamics data from CML patient 0124
author: Stephanie Hicks, Yuchao Jiang
output: 
    html_document:
        theme: cosmo 
        toc: true
        toc_float: true
        highlight: tango
        number_sections: false
fig_width: 5
fig_height: 5 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Summary of patient data

This is scRNA-seq (indrop platform) data from a MPN patient 
(denoted 0124 patient). The data is stored in a MATLAB (`.MAT`) 
format. There are the following data sets included in the object: 

* `E` = normalized gene expression for all the cells (5126 genes, 25419 cells)
* `cellType` = vector of length 5126 representing the cancer status for each cell. 0 means the cell type is not known. 1 means a wild-type cell. 2 means a cancer cell.
* `count_matrix` = raw counts (5126 genes, 25419 cells)
* `gene.filter` = genes that passed previous quality control filter from SPRING. If a gene is 1 it passed, 0 it did not.
* `gene_list` = name of the all genes in the same orders as the columns of the `E` matrix and the `count_matrix`


## Previous analyses reported by Sahand

_We use SPRING, a dimensionality reduction tool that uses force-directed graph to plot the differentiation trajectories. SPRING has some criteria for selecting a subset of genes that are informative (for example fano factor and mean expression thresholds). There are 353 genes that pass this filtering. These genes are stored in "gene_filter" vector. If a gene is 1 it passed, 0 it did not. SPRING uses the reduced 5126 by 353 gene expression matrix as its starting point (then does PCA to 20 dimensions). tSNE also works well on this data._

# Biological question of interest

* Interested in identifying genes that are over-expressed or under-expressed in the cancer cells.
* Are there differences between cancer and normal cells in genes that regulate cell cycle?
* Hypothesis: Previous analyses suggested the gene E2F4 is over expressed in the cancer cells.


# Data Analysis

Immediate to-dos: 

* identify doublets, empty cells, and and apply normalization
* Use `scran::cyclone()` in Bioconductor to compare cancer vs normal cells in different stages of cell cycle


## Load data
Read in data 
```{r, message=FALSE, cache=TRUE}
library(R.matlab)
dat <- readMat("data/0124Patient_Data.mat") # dat is 75 MB in size

sce_colData <- data.frame("cancer_status" = dat$cellType)
sce_rowData <- data.frame("gene_name" = as.character(unlist(dat$gene.list)), 
                          "gene_filter_spring" = c(dat$gene.filter))
sce_rowData$gene_name <- as.character(sce_rowData$gene_name)
rownames(sce_rowData) <- sce_rowData$gene_name
```

Create a `SingleCellExperiment` object
```{r, eval=FALSE}
library(SingleCellExperiment)
library(scater)
sce <- SingleCellExperiment(assays = list(counts = t(dat$count.matrix)), 
          rowData = sce_rowData, colData = sce_colData)

rowData(sce)$gene_name_clean <- gsub(pattern = "\\.[0-9]+", replacement = "", x = rownames(sce))
sce <- sce[!duplicated(rowData(sce)$gene_name_clean), ] 
rownames(sce) <- rowData(sce)$gene_name_clean

sce <- scater::getBMFeatureAnnos(sce,
    filters = "hgnc_symbol",
    attributes = c("ensembl_gene_id",
        "hgnc_symbol", "chromosome_name",
        "start_position", "end_position"),
    feature_symbol = "hgnc_symbol",
    feature_id = "ensembl_gene_id",
    biomart = "ENSEMBL_MART_ENSEMBL",
    dataset = "hsapiens_gene_ensembl",
    host = "www.ensembl.org")

# separate information about mitochondrial genes
isSpike(sce, "MT") <- grepl("^MT-", rownames(sce)) 

# calculate QC metrics
sce <- calculateQCMetrics(sce, 
          feature_controls = list(MT = isSpike(sce, "MT")))

# save SCE object
saveRDS(sce, file = "data/sce.rds")
```

## EDA and QC on the cells

```{r}
library(SingleCellExperiment)
library(scater)
sce <- readRDS("data/sce.rds")
```


Histogram of library sizes and number 
of expressed genes.
```{r}
par(mfrow=c(1,2))
hist(sce$total_counts/1e4, xlab="Library sizes (thousands)", main="",
     breaks=20, col="grey80", ylab="Number of cells")
hist(sce$total_features, xlab="Number of expressed genes", main="",
     breaks=20, col="grey80", ylab="Number of cells")
```


```{r}
boxplot(sce$total_counts ~ sce$cancer_status)
boxplot(sce$total_features_by_counts ~ sce$cancer_status)
boxplot(sce$pct_counts_MT ~ sce$cancer_status)

plot(sce$total_features_by_counts, sce$pct_counts_MT)
```
There are definitely cells that have much higher pct MT
that need to be removed. 


### Remove low-quality cells 

We remove cells with log-library sizes that are more than 
3 median absolute deviations (MADs) below the median 
log-library size. (A log-transformation improves resolution 
at small values, especially when the MAD of the raw values 
is comparable to or greater than the median). Next, we 
remove cells where the log-transformed number of expressed 
genes is 3 MADs below the median. Then, we remove cells 
with percent mitochondria 3 MADs above the median. 

```{r}
libsize.drop <- isOutlier(sce$total_counts, nmads=3, type="lower", log=TRUE)
feature.drop <- isOutlier(sce$total_features, nmads=3, type="lower", log=TRUE)
MT.drop <- isOutlier(sce$pct_counts_MT, nmads=3, type="higher", log=TRUE)

sce <- sce[,!(libsize.drop | feature.drop | MT.drop)]
# sanity check
plot(sce2$total_features_by_counts, sce2$pct_counts_MT)

```


### Classification of cell cycle phase

We use the prediction method described by Scialdone 
et al. (2015) to classify cells into cell cycle phases 
based on the gene expression data. Using a training dataset, 
the sign of the difference in expression between two genes 
was computed for each pair of genes. Pairs with changes in the 
sign across cell cycle phases were chosen as markers. Cells in 
a test dataset can then be classified into the appropriate 
phase, based on whether the observed sign for each marker pair 
is consistent with one phase or another.

This approach is implemented in the cyclone function from the 
scran package. The package contains a pre-trained set of marker 
pairs for mouse and human data, which we can load in the the 
`readRDS()` function. We use the Ensembl identifiers for each 
gene in our dataset to match up with the names in the pre-trained 
set of gene pairs.

This can take a while to run. 
```{r}
library(scran)
human.pairs <- readRDS(system.file("exdata", "human_cycle_markers.rds", package="scran"))
assigned <- cyclone(sce, pairs=human.pairs, gene.names=rowData(sce)$ensembl_gene_id)
head(assigned$scores)

plot(assigned$score$G1, assigned$score$G2M, 
    xlab="G1 score", ylab="G2/M score", pch=16)
```
The cyclone result for each cell in the dataset is shown above. 
Each cell is assigned a score for each phase, with a higher score 
corresponding to a higher probability that the cell is in that 
phase. We focus on the G1 and G2/M scores as these are the 
most informative for classification.

Cells are classified as being in G1 phase if the G1 score 
is above 0.5 and greater than the G2/M score; in G2/M phase 
if the G2/M score is above 0.5 and greater than the G1 score; 
and in S phase if neither score is above 0.5. We save these 
assignments into the `SingleCellExperiment` object for later use.

```{r}
sce$phases <- assigned$phases
table(sce$phases)

x <- table(sce$cancer_status[sce$cancer_status != 0], 
      sce$phases[sce$cancer_status != 0])  
x / rowSums(x)

chisq.test(x)
```

**Note**: We do not filter out low-abundance genes before 
applying cyclone. Even if a gene is not expressed in any cell, 
it may still be useful for classification if it is phase-specific. 
Its lack of expression relative to other genes will still yield 
informative pairs, and filtering them out would reduce power.

## EDA and QC on the genes

```{r}
# Remove genes not expressed in at least 5 cells
keep_feature <- rowSums(counts(sce) > 0) > 5
table(keep_feature)
sce <- sce[keep_feature, ]
```



# Run Seurat with additional QC

Below, we create a Seurat object and apply additional qualtiy control procedures on both cells and genes.

```{r}
library(Seurat)
CML0124_raw=counts(sce)
colnames(CML0124_raw)=paste('cell',1:ncol(CML0124_raw),sep='')
CML0124 <- CreateSeuratObject(raw.data = CML0124_raw, min.cells = 3, min.genes = 200, 
                            project = "CML0124")

mito.genes <- grep(pattern = "^MT-", x = rownames(x = CML0124@data), value = TRUE)
percent.mito <- Matrix::colSums(CML0124@raw.data[mito.genes, ])/Matrix::colSums(CML0124@raw.data)

CML0124 <- AddMetaData(object = CML0124, metadata = percent.mito, col.name = "percent.mito")
VlnPlot(object = CML0124, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)

par(mfrow=c(1,2))
nUMI=apply(CML0124_raw,2,sum)
nGene=apply(CML0124_raw,2,function(x){sum(x>0)})
hist(nUMI,100, xlab='Number of UMIs',main='Number of UMIs detected per cell')
hist(nGene,100, xlab='Number of genes',main='Number of genes detected per cell')

# Cell QC: remove cells with 3 MAD below or above the median
filter1= (nUMI >= median(nUMI)-3*mad(nUMI)) & (nUMI <= median(nUMI)+3*mad(nUMI))
filter2= (nGene >= median(nGene)-3*mad(nGene)) & (nGene <= median(nGene)+3*mad(nGene))
CML0124_raw=CML0124_raw[,filter1&filter2]
sce=sce[,filter1&filter2]

# Cell QC: remove cells with greater than 20% of UMIs mapped to mitochondria:
mito.prop=apply(CML0124_raw[grep('^MT-',rownames(CML0124_raw)),],2,sum)/apply(CML0124_raw,2,sum)
par(mfrow=c(1,1))
hist(mito.prop,100)
filter=mito.prop<=0.2
CML0124_raw=CML0124_raw[,filter]
sce=sce[,filter]

# Cell QC: look at ratio of nUMI and nGene
nUMI=apply(CML0124_raw,2,sum)
nGene=apply(CML0124_raw,2,function(x){sum(x>0)})
plot(nUMI, nGene)
spl=smooth.spline(nGene~nUMI)
points(spl[[1]],spl[[2]],col=2,type='l',lwd=2)
resid=(nGene-predict(spl,nUMI)$y)/nGene
hist(resid,100)

plot(nUMI, nGene)
points(spl[[1]],spl[[2]],col=2,type='l',lwd=2)
filter= (resid >= median(resid)-3*mad(resid)) & (resid <= median(resid)+3*mad(resid))
points(nUMI[!filter], nGene[!filter],col=3, pch=16)
CML0124_raw=CML0124_raw[,filter]
sce=sce[,filter]

# Gene QC: 1 CPM in at least 5 cells
N=apply(CML0124_raw,2,sum)/10^6
CPM=round(t(apply(CML0124_raw,1,function(x){x/N})),2)
filter= apply(CPM,1,function(x){sum(x>=1)>=5})
CML0124_raw=CML0124_raw[filter,]
CPM=CPM[filter,]
sce=sce[filter,]
dim(sce); dim(CPM); dim(CML0124_raw)
```

Below is an application of Seurat after QC.

```{r}
# create seurat object and calculate mito proportion after QC
CML0124 <- CreateSeuratObject(raw.data = CML0124_raw, min.cells = 3, min.genes = 200, 
                            project = "CML0124")
mito.genes <- grep(pattern = "^MT-", x = rownames(x = CML0124@data), value = TRUE)
percent.mito <- Matrix::colSums(CML0124@raw.data[mito.genes, ])/Matrix::colSums(CML0124@raw.data)
CML0124 <- AddMetaData(object = CML0124, metadata = percent.mito, col.name = "percent.mito")
VlnPlot(object = CML0124, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)
CML0124@meta.data=cbind(CML0124@meta.data, cancer_status=sce@colData$cancer_status)

# Filter cells
CML0124 <- FilterCells(object = CML0124, subset.names = c("nGene", "percent.mito"), 
                     low.thresholds = c(200, -Inf), high.thresholds = c(Inf, 1))

# log normalize data
CML0124 <- NormalizeData(object = CML0124, normalization.method = "LogNormalize", 
                       scale.factor = 10000)

# Find highly variable genes
par(mfrow=c(1,1))
CML0124 <- FindVariableGenes(object = CML0124, mean.function = ExpMean, dispersion.function = LogVMR, 
                           x.low.cutoff = 0.1, x.high.cutoff = 6, y.cutoff = 0.5)
head(CML0124@var.genes)
spring.var.genes=rownames(CML0124_raw)[which(rowData(sce)$gene_filter_spring==1)]
intersect(CML0124@var.genes, spring.var.genes)

# Scale data and remove unwanted sources of variation: this is based on linear scaling
CML0124 <- ScaleData(object = CML0124, vars.to.regress = c("nUMI", "percent.mito"))

# PCA
CML0124 <- RunPCA(object = CML0124, pc.genes = CML0124@var.genes, do.print = TRUE, pcs.print = 1:5, 
                genes.print = 5, pcs.compute = 40)
VizPCA(object = CML0124, pcs.use = 1:2)
PCAPlot(object = CML0124, dim.1 = 1, dim.2 = 2)

# Determine the number of PCs
PCElbowPlot(object = CML0124, num.pc = 40)

# Cluster cells
CML0124 <- FindClusters(object = CML0124, reduction.type = "pca", dims.use = 1:20, 
                      resolution = 0.8, print.output = 0, save.SNN = TRUE, force.recalc = TRUE)
PrintFindClustersParams(object = CML0124)

# Run t-SNE
CML0124 <- RunTSNE(object = CML0124, dims.use = 1:20, do.fast = TRUE)

TSNEPlot(object = CML0124, do.label = T)
TSNEPlot(object = CML0124, group.by = 'cancer_status')

```

Perform differential expression analysis and calculate spearman correlation between each pair of clusters.
```{r}

# Find number of differentially expressed genes between each pair of clusters
nClusters = length(unique(CML0124@ident))
clusterQcMat = matrix(0, nClusters, nClusters)
for ( i in 1:nClusters ) {
  for ( j in 1:nClusters ) {
    if (i < j){
      markers = FindMarkers(object = CML0124, ident.1 = i-1 , ident.2 = j-1 , min.pct = 0.25)
      clusterQcMat[ i, j ] = nrow(markers[markers$p_val_adj < .01,])
    }
  }
}

clusterQcMatSym = clusterQcMat + t(clusterQcMat)
rownames(clusterQcMatSym) = paste(0:(nClusters - 1))
colnames(clusterQcMatSym) = paste(0:(nClusters - 1))
save(clusterQcMatSym, file='data/clusterQcMatSym.rda')

library(pheatmap)
temp=pheatmap(clusterQcMatSym)
pheatmap(clusterQcMatSym[temp$tree_row$order,temp$tree_row$order],cluster_rows = F, cluster_cols = F)

# Construct pseudo bulk for each cluster, find pair-wise similarity (correlation) between each pair of clusters
pseudo.bulk=matrix(nrow=nrow(CML0124_raw), ncol=nClusters)
for(i in 1:nClusters){
  cat(i,'\n')
  pseudo.bulk[,i]=apply(CML0124_raw[,which(CML0124@ident==(i-1))],1,sum)
}
colnames(pseudo.bulk)=0:(nClusters-1)
rownames(pseudo.bulk)=rownames(CML0124_raw)
head(pseudo.bulk)
pheatmap(cor(pseudo.bulk, method='spearman'))
pheatmap(cor(pseudo.bulk, method = 'spearman')[temp$tree_row$order,temp$tree_row$order],cluster_rows = F, cluster_cols = F)
```
