---
title: Analysis of Single Cell Dynamics data from CML patient 0124
author: Stephanie Hicks
output: 
    html_document:
        theme: cosmo 
        toc: true
        toc_float: true
        highlight: tango
        number_sections: false
fig_width: 5
fig_height: 5 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Summary of patient data

This is scRNA-seq (indrop platform) data from a MPN patient 
(denoted 0124 patient). The data is stored in a MATLAB (`.MAT`) 
format. There are the following data sets included in the object: 

* `E` = normalized gene expression for all the cells (5126 genes, 25419 cells)
* `cellType` = vector of length 5126 representing the cancer status for each cell. 0 means the cell type is not known. 1 means a wild-type cell. 2 means a cancer cell.
* `count_matrix` = raw counts (5126 genes, 25419 cells)
* `gene.filter` = genes that passed previous quality control filter from SPRING. If a gene is 1 it passed, 0 it did not.
* `gene_list` = name of the all genes in the same orders as the columns of the `E` matrix and the `count_matrix`


## Previous analyses reported by Sahand

_We use SPRING, a dimensionality reduction tool that uses force-directed graph to plot the differentiation trajectories. SPRING has some criteria for selecting a subset of genes that are informative (for example fano factor and mean expression thresholds). There are 353 genes that pass this filtering. These genes are stored in "gene_filter" vector. If a gene is 1 it passed, 0 it did not. SPRING uses the reduced 5126 by 353 gene expression matrix as its starting point (then does PCA to 20 dimensions). tSNE also works well on this data._

# Biological question of interest

Interested in identifying genes that are over-expressed or under-expressed in the cancer cells. Genes that regulate cell cycle would be a good place to start. We think the gene E2F4 (column 6114 of the E matrix) is over expressed in the cancer cells.


# Data Analysis

## Load data
Read in data 
```{r, message=FALSE, cache=TRUE}
library(R.matlab)
dat <- readMat("data/0124Patient_Data.mat") # dat is 75 MB in size

sce_colData <- data.frame("cancer_status" = dat$cellType)
sce_rowData <- data.frame("gene_name" = as.character(unlist(dat$gene.list)), 
                          "gene_filter_spring" = c(dat$gene.filter))
rownames(sce_rowData) <- sce_rowData$gene_name
```

Create a `SingleCellExperiment` object
```{r, eval=FALSE}
library(SingleCellExperiment)
library(scater)
sce <- SingleCellExperiment(assays = list(counts = t(dat$count.matrix)), 
          rowData = sce_rowData, colData = sce_colData)

# Remove genes not expressed in at least 5 cells
keep_feature <- rowSums(counts(sce) > 0) > 5
table(keep_feature)
sce <- sce[keep_feature, ]

# calcualte QC metrics
sce <- calculateQCMetrics(sce)

# save SCE object
saveRDS(sce, file = "data/sce.rds")
```

## Exploratory Data Analysis

Histogram of library sizes and number 
of expressed genes.
```{r}
par(mfrow=c(1,2))
hist(sce$total_counts/1e4, xlab="Library sizes (thousands)", main="",
     breaks=20, col="grey80", ylab="Number of cells")
hist(sce$total_features, xlab="Number of expressed genes", main="",
     breaks=20, col="grey80", ylab="Number of cells")
```

We remove cells with log-library sizes that are more than 3 median absolute deviations (MADs) below the median log-library size. (A log-transformation improves resolution at small values, especially when the MAD of the raw values is comparable to or greater than the median.) We also remove cells where the log-transformed number of expressed genes is 3 MADs below the median.
```{r}
libsize.drop <- isOutlier(sce$total_counts, nmads=3, type="lower", log=TRUE)
feature.drop <- isOutlier(sce$total_features, nmads=3, type="lower", log=TRUE)

sce <- sce[,!(libsize.drop | feature.drop)]
```




